DataProcessor.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * @license GNU GPL v3+
+ */
+
+namespace Piwik\Plugins\CampaignVisitorsByTime;
+
+use Piwik\Site;
+
+/**
+ * Database access functionality for plugin CampaignVisitorsByTime
+ */
+class DataProcessor {
+
+	private $campaigns = [];
+	private $keywords = [];
+
+	public function processDatasets( \Zend_Db_Statement $dataSets ): void {
+		while ( $row = $dataSets->fetch() ) {
+
+			$campaign = $row['referer_name'];
+			$keyword = $row['referer_keyword'];
+			$timestamp = $row['timestamp_floored'];
+
+			if ( !array_key_exists( $campaign, $this->campaigns ) ) {
+				$this->initCampaign( $campaign );
+			}
+
+			if ( !array_key_exists( $keyword, $this->keywords[$campaign] ) ) {
+				$this->initKeywordForCampaign( $campaign, $keyword );
+			}
+
+			$timestampFloored = $this->parseTimestamp(
+				$timestamp,
+				$this->getTimeZoneForSite( $row['idsite'] )
```

Comment:
You mentioned that getting the timezone involves querying the database. Since this is called for *every* log row I think it would be a great speedup if you cached the results inside this class (using a private array).

Reply:
Don't know how I missed that myself, good idea!